lib/vector-core/src/event/finalization.rs

```diff
@@ -0,0 +1,340 @@
+    fn drop(&mut self) {
+        self.update_batch();
+    }
+}
+
+/// A convenience type alias for the one-shot receiver for an individual
+/// batch status.
+pub type BatchStatusReceiver = oneshot::Receiver<BatchStatus>;
+
+/// A batch notifier contains the status of the current batch along with
+/// a one-shot notifier to send that status back to the source. It is
+/// shared among all events of a batch.
+#[derive(Debug)]
+pub struct BatchNotifier {
+    status: Atomic<BatchStatus>,
+    notifier: Option<oneshot::Sender<BatchStatus>>,
+}
+
+impl BatchNotifier {
+    /// Create a new `BatchNotifier` along with the receiver used to
+    /// await its finalization status.
+    pub fn new_with_receiver() -> (Arc<Self>, BatchStatusReceiver) {
+        let (sender, receiver) = oneshot::channel();
+        let notifier = Self {
+            status: Atomic::new(BatchStatus::Delivered),
+            notifier: Some(sender),
+        };
+        (Arc::new(notifier), receiver)
+    }
+
+    /// Update this notifier's status from the status of a finalized event.
+    fn update_status(&self, status: EventStatus) {
+        // The status starts as Delivered and can only change to Failed
+        // here. A store cycle is much faster than fetch+update.
+        if status == EventStatus::Failed {
+            self.status.store(BatchStatus::Failed, Ordering::Relaxed);
+        }
+    }
+
+    /// Send this notifier's status up to the source.
+    fn send_status(&mut self) {
+        if let Some(notifier) = self.notifier.take() {
+            let status = self.status.load(Ordering::Relaxed);
+            // Ignore the error case, as it will happen during normal
+            // source shutdown and we can't detect that here.
+            let _ = notifier.send(status);
+        }
+    }
+}
+
+impl Drop for BatchNotifier {
+    fn drop(&mut self) {
+        self.send_status();
+    }
+}
+
+/// The status of an individual batch as a whole.
+#[derive(Atom, Copy, Clone, Debug, Derivative, Deserialize, Eq, PartialEq, Serialize)]
+#[derivative(Default)]
+#[repr(u8)]
+pub enum BatchStatus {
+    /// All events in the batch were accepted (the default)
+    #[derivative(Default)]
+    Delivered,
+    /// At least one event in the batch failed delivery.
+    Failed,
+}
+
+// Can be dropped when this issue is closed:
+// https://github.com/LukasKalbertodt/atomig/issues/3
+impl AtomInteger for BatchStatus {}
+
+/// The status of an individual event.
+#[derive(Atom, Copy, Clone, Debug, Derivative, Deserialize, Eq, PartialEq, Serialize)]
+#[derivative(Default)]
+#[repr(u8)]
+pub enum EventStatus {
+    /// All copies of this event were dropped without being finalized (the default).
+    #[derivative(Default)]
+    Dropped,
+    /// All copies of this event were delivered successfully.
+    Delivered,
+    /// At least one copy of this event failed to be delivered.
+    Failed,
+    /// This status has been recorded and should not be updated.
+    NoOp,
```

Comment:
Would it make sense to name this `Recorded` to better match the others? Not a big deal either way, just an idea.

Reply:
Makes sense